### DISCUSSION TITLE
FFMPEG Transcode - multiple streams

### DISCUSSION THREAD

**Discussion by jinfanzon** [k-yle/rtsp-relay#69] [Q&A]
Hi - I'm trying to implement this module to work with multiple backend rtsp cams. My main concern is if the transcode occurs at frontend side (client) or in the backend.

**Comment by baur** [comment]
Hi

[Usage with many cameras](https://github.com/k-yle/rtsp-relay#usage-with-many-cameras)

It helped me!
Now the app shows 20 cameras simultaneously

**Reply by jinfanzon** [reply]
Hi @baur  - Thanks for the reply. Actually, I was able to make it work with several cams and that approach. But my concern is if the transcode is done in the backend or on the client side.

**Reply by baur** [reply]
after all, rtsp-relay this is precisely for the backend transcoding. for the client-side, we are using JS MJPEG Player witch play from Web Socket

**Reply by jinfanzon** [reply]
Thanks @baur - Is there any performance metric or cpu usage statistics? It'd be very helpful to understand the resource usage of this module per stream.

**Reply by baur** [reply]
CPU 2.5 GHz. 4 Virtual core, 2 Sockets 
For the first stream was 8-10% per camera, for the second stream was 1-3% per camera

If less than 5 camera recommend http://umediaserver.net
Free version limitations: 5 live broadcasts can be configured, 10 concurrent player connections can be served.

**Reply by jinfanzon** [reply]
Hi Baur! I run some new checks with the option additionalFlags: ['-q', '1'] and cpu usage escalated up to 60%. is that possible?

**Reply by jinfanzon** [reply]
Hi @baur  - I wanted to understand if this tool is transcoding or encoding.

**Reply by baur** [reply]
> additionalFlags

I didn't have this, I probably misused flags

**Reply by baur** [reply]
> Hi @baur - I wanted to understand if this tool is transcoding or encoding.

To be honest, I don't know the difference

It seems that rtsp-relay converts an RTSP to MJPEG and starts broadcasting over sockets, our task is to play this on the client-side through any JS-player

**Reply by k-yle** [reply]
> I wanted to understand if this tool is transcoding or encoding.

this library uses ffmpeg to transcode the RTSP stream into MPEG1. the MPEG1 stream is sent to the client via websockets. jsmpeg uses an HTML Canvas element to render the MPEG1 stream. 

> I run some new checks with the option additionalFlags: ['-q', '1'] and cpu usage escalated up to 60%. is that possible?

That is expected. If you want a higher quality image, more bandwidth is required and also more data needs to be transcoded per second. The `-q` flag refers to ffmpeg's qscale option. [some info here](https://stackoverflow.com/a/10234065/5470183) on what the values mean. If `-q 1` uses too much resources or bandwith, you could increase it a bit (`1` = best quality, `31` = worse quality)

**Reply by jinfanzon** [reply]
Would it be possible/feasible to transmux instead of transcode so as to save cpu usage?

**Reply by k-yle** [reply]
@jinfanzon [these are the current arguments](https://github.com/k-yle/rtsp-relay/blob/a137bb4/index.js#L32-L41) to ffmpeg, if you can figure out a more efficient method, feel free to make a PR (or just let me know)
